impactlab_tools/utils/configdict.py

```diff
@@ -0,0 +1,163 @@
+"""Class for representing tool configuration files
+"""
+
+import collections
+import inspect
+
+
+def gather_configtree(d):
+    """Chains nested-dicts into a connected tree of ConfigDict(s)
+
+    Parameters
+    ----------
+    d : dict
+        Cast to :py:class:`ConfigDict`. Nested dicts within are also
+        recursively cast and assigned parents, reflecting their nested
+        structure.
+
+    Returns
+    -------
+    out : ConfigDict
+
+    Examples
+    --------
+    .. code-block:: python
+
+        >>> nest = {'a': 1, 'b': {'a': 2}, 'c': 3, 'd-4': 4, 'e_5': 5, 'F': 6}
+        >>> tree = gather_configtree(nest)
+        >>> tree['b']['d-4']
+        4
+    """
+    out = ConfigDict(d)
+    for k, v in out.data.items():
+        # Replace nested maps with new ConfigDicts
+        if isinstance(v, collections.abc.MutableMapping):
+            out.data[k] = gather_configtree(v)
+            out.data[k].parent = out
+    return out
+
+
+class ConfigDict(collections.UserDict):
+    """Chain-able dictionary to hold projection configurations.
+
+    A ConfigDict is a dictionary-like interface to a chainmap/linked list.
+    Nested dicts can be access like a traditional dictionary but it searches
+    parent dictionaries for keys:values not found. All string keys normalized,
+    by transforming all characters to lowercase, and all underscores to
+    hyphens.
+
+    Attributes
+    ----------
+    parent : ConfigDict or None
+        Parent ConfigDict object to query for keys if not in `self.data`.
+    key_access_stack : dict
+        Dictionary with values giving the :py:func:`inspect.stack()` from the
+        most recent time a key was retrieved (via `self.__getitem__()`).
+    data : dict
+        The 'local' dictionary, not in parents.
+
+    See Also
+    --------
+    gather_configtree : Chains nested-dicts into a connected tree of
+        ConfigDict(s)
+
+    Examples
+    --------
+    .. code-block:: python
+
+        >>> d = {'a': 1, 'b': {'a': 2}, 'c': 3, 'd-4': 4, 'e_5': 5, 'F': 6}
+        >>> cd = ConfigDict(d)
+        >>> cd['b']
+        {'a': 2}
+
+        'F' key is now lowercase.
+
+        >>> cd['f']
+        6
+
+        '_' is now '-'
+
+        >>> cd['e-5']
+        5
+
+        Keys that have been accessed.
+
+        >>> cd.key_access_stack.keys()
+        dict_keys(['b', 'f', 'e-5'])
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.parent = None
+
+        for k, v in self.data.items():
+            # Normalize string keys, only when needed
+            if isinstance(k, str) and (k.isupper() or '_' in k):
+                new_key = self._normalize_key(k)
+                self.data[new_key] = self.pop(k)
```

Comment:
I think you actually want to wait to do this until key-access. If you add them here (as I do when I'm just working with dicts), you will end up keeping track of accesses to the two keys separately (e.g., what happens if one part of the code accesses e-5 and the other access E_5?).

Reply:
I see that you do it also on access, and remove the original. So my objection above is not a problem, but it seems less efficient than just doing it on access.

Reply:
The key access isn't logged during `__init__()` because we query the local dictionary `self.data` directly.

I figured that normalizing strings and storing them once during initialization would be more efficient that having to normalize them every time we climb the config tree and search for matching keys.

That said, I don't know the exact use cases like you do so I might eat my hat on this. :-)